doc/design/record_fetcher_proposal.md

```diff
@@ -0,0 +1,170 @@
+lot of structure for this, as any relationship between documents will be encoded by the manifests -- the service just
+needs to be able to store the raw documents for later lookup (but lookups don't need to be via this service). Backing
+this storage service with a fairly flat filesystem is fine here, we're rarely going to list files, only write and read.
+
+A big benefit of content-addressed storage here is that we get deduping almost for free -- we may need to zero out
+content that changes on every page load (e.g. "This page requested at HH:MM"), but that's more of an optimization than a
+hard requrement. These savings would apply to both documents which remain unchanged between fetches and documents that
+are shared between records.
+
+### Manifest Construction
+
+Manifests need to be signed, so we'll just use [RFC 7519](https://tools.ietf.org/html/rfc7519) JWT blobs, encoding the
+manifest's information in the JWT payload. As an example, the payload could resemble:
+
+```json
+{
+  "version": 1,
+  "uuid": "6e566397-6803-49ec-9280-62c00b368e3f",
+  "record_type": "USA/FL/Bay/Court",
+  "case_id": "20000001",
+  "collected_at_ms": 1592618026000,
+  "fetcher_id": "github.com/Police-Data-Accessibility-Project/Scrapers:USA/FL/Bay/Court/scraper/Scraper.py@1db20ce1e63768363d532d46a641722986f63524",
+  "documents": [
+    {
+      "label": "some pdf we'll have to ocr",
+      "content_type": "application/pdf",
+      "digest": "...",
+    }, {
+      "label": "party information",
+      "content_type": "text/html",
+      "digest": "...",
+    }
+  ]
+}	
+```
+
+### Record Bundles
+
+Signed manifests and all related documents will be uploaded together as [MIME Multipart messages](
+https://en.wikipedia.org/wiki/MIME#Multipart_messages). Authentication will be via a JWT bearer token (how this is
+granted is not in scope). This has a number of benefits, but the largest is that it gives us a single point to
+validate the manifest and documents together before changing any state. This will allow us to:
+1. verify the signature of the manifest against the bearer auth token to ensure they're coming from the same entity
+2. ensure no documents are missing from the bundle by verifying a manifest's digests against the documents' contents
+3. ensure the manifest is complete by verifying that no additional documents have been uploaded
+4. ensure all required fields are present
+
+If any of those conditions are not met, the request is deemed invalid.
+
+## Caveats
+This is somewhat complex and may be tricky to implement. A way to mitigate this is to invest in easy-to-use canonical
+implementations in common languages with an API that makes misuse hard. For example, a Python API could resemble:
+
+```python
+with Record(record_type="USA/FL/Bay/Court", case_id="1337") as r:
+  r.AddHtml("label1", html_file1)
+  r.AddHtml("label2", html_file2)
+  r.AddPdf("label3", pdf_file)
+  r.commit()  # construct a manifest, upload bundle of manifest + documents
```

Comment:
I like how fluent and readable this is.

Typically in a builder pattern you'd expect the builder to return itself. So you could do:

r
  .AddHtml("label1", html_file1)
  .AddHtml("label2", html_file2)
  .AddPdf("label3", pdf_file)
  .commit()

I'll be honest, I'm no python expert. So maybe that's not super pythonic, but regardless of the builder returning it self or just calling on "r" each time, this is super readable and I just thought it was worth while to point that out.

Reply:
Java really lends itself to that kind of expressions, but we'd have to keep Python from ending the expression prematurely. Something like:
```python
r\
.addHtml("label1", html_file1)\
.addHtml("label2", html_file2)\
.commit()
```
Which is kinda gross, or:
```python
(r
).addHtml("foo", bar
).commit()
```
... which is absolutely cursed.

Anyway, thanks for the feedback! I'm a big fan of try to make hard things simple. 😄